### PR TITLE
MAP 102 - Create Prototype GetCapabilities Request

### PR DESCRIPTION
**Pull request by wchen329**
The GetCapabilities Request is mostly done; the XML generated by ncWFS opens successfully in QGIS (it shows the only feature hru_soil_moist with title "HRU Soil Moisture").

- GetCapabilities is mostly static at this point though contact information will change according to what is in threddsConfig.xml.
- Adding the layer itself doesn't work yet in QGIS. Need to implement DescribeFeature and GetFeature for full functionality.
- Only WFS 2.0 is reported as accepted.

And lastly and most importantly, this PR also includes the WFS spring controller for the new WFS Servlet.

### CODE REVIEW

File: tds/src/main/java/thredds/server/wfs/WFSController.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+			String request = hsreq.getParameter("REQUEST");
+			
+			// 
+			if(request != null) {
+				if(request.equals(WFSRequestType.GetCapabilities.toString())) {
+					getCapabilities(wr, hsreq);
+				}
+			}
+			
+			else{
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/WFSController.java:62]: Could an HTTP 404 be returned here? Something along the lines of `hsres.sendError(hsres.SC_NOT_FOUND, "Not Found - ncWFS on THREDDS");` ?

File: tds/src/main/java/thredds/server/wfs/WFSRequestType.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
+package thredds.server.wfs;
+
+/**
+ * Enum representing WFS Request Types as specified by the OGC.
+ * 
+ * @author wchen@usgs.gov
+ *
+ */
+public enum WFSRequestType {
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/WFSRequestType.java:9]: Good call to use an enum for this!

File: tds/src/main/java/thredds/server/wfs/WFSController.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+				wr.append("ncWFS on THREDDS - ");
+				wr.append(hsreq.getScheme() + "://" + hsreq.getServerName() + ":" + hsreq.getServerPort() + "/thredds/wfs");
+			}
+		}
+		
+		catch(IOException io) {
+
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/WFSController.java]: Should this catch be empty? Alternatives include:
 * throwing a new unchecked exception that wraps the IOException. Example: `throw new RuntimeException("my custom message", io);`
 * logging the exception
 * printing the exception message to stderr

File: tds/src/main/java/thredds/server/wfs/WFSGetCapabilitiesWriter.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+import thredds.server.config.ThreddsConfig;
+
+/**
+ * A writer for a WFS compliant Geospatial XML given a print writer.
+ * Specifically writes the XML for WFS GetCapabilities requests.
+ * 
+ * @author wchen@usgs.gov
+ *
+ */
+public class WFSGetCapabilitiesWriter {
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/WFSGetCapabilitiesWriter.java:15]: I like that you put the GetCapabilities functionality in this class rather than in the WFSController. This is an excellent separation of concerns!

File: tds/src/main/java/thredds/server/wfs/WFSFeature.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+import org.geotoolkit.referencing.CRS;
+
+/**
+ * A simple container for a WFS Feature Type
+ * 
+ * @author wchen@usgs.gov
+ *
+ */
+public class WFSFeature {
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/WFSFeature.java:9]: I like this little class! I think it will be a good landing place for information from THREDDS' catalog in the future!

File: tds/src/main/java/thredds/server/wfs/WFSFeature.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
+package thredds.server.wfs;
+
+import org.geotoolkit.referencing.CRS;
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/WFSFeature.java]: Is this import used?

File: tds/src/main/java/thredds/server/wfs/WFSFeature.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * A simple container for a WFS Feature Type
+ * 
+ * @author wchen@usgs.gov
+ *
+ */
+public class WFSFeature {
+	private String name;
+	private String title;
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/WFSFeature.java]: I notice that neither of these instance variables are modified after they are set in the constructor. Are you aiming to make this class immutable? If so, would it be appropriate to add the `final` modifier to the instance variables?

File: tds/src/main/java/thredds/server/wfs/WFSGetCapabilitiesWriter.java (hunk 4 above)
Comment (cschroedl-gov) [reply tds/src/main/java/thredds/server/wfs/WFSGetCapabilitiesWriter.java:15]: I like the granularity and naming of these methods.

Comment (cschroedl-gov) [reply tds/src/main/java/thredds/server/wfs/WFSGetCapabilitiesWriter.java:15]: This class is highly stateful. Sometimes stateful designs are appropriate. They are usually harder to test than stateless designs. Given our goals and time constraints I do not want to mandate tests on this class, however I would like to check: have you considered a stateless or a less stateful design?

DZone has [a good article](https://dzone.com/articles/stateful-or-stateless-classes) on this topic if you are hungry for inspiration.

A few examples of state and possible alternatives:
 * `fileOutput`, a String that builds up the XML to be written to the `PrintWriter`. This could be removed in favor of writing results to the `PrintWriter` immediately, or by returning the results of each method as a String.
 * a list of WFSRequestTypes. This could be changed from an instance variable to a parameter on methods.
 * a list of WFSFeatures. This could be changed from an instance variable to a parameter on methods.

Just because these things could be done doesn't mean they should be done.  If you have weighed the pros and cons and feel a stateful design is appropriate, go for it. If you think a stateful design would be useful, but you would rather implement it in a subsequent PR, that is OK too.

**Review by cschroedl-gov — CHANGES_REQUESTED** [review]
Good separation of concerns! Good method names! I'm most eager to see how you would address the comments I made in the `httpHandler` method. The other stuff is just icing on the cake.

File: tds/src/main/java/thredds/server/wfs/WFSController.java (hunk 1 above)
Comment (wchen329) [reply tds/src/main/java/thredds/server/wfs/WFSController.java:62]: Agreed, yeah the default message was kind of a funny little thing will be better to make it that HTTP 404 (though wfs servers seem to implement this ows:Exception i.e. http://demo.mapserver.org/cgi-bin/wfs?VERSION=2.0.0&SERVICE=wfs so should 404 be wrapper inside one of these?)!

File: tds/src/main/java/thredds/server/wfs/WFSRequestType.java (hunk 2 above)
Comment (wchen329) [reply tds/src/main/java/thredds/server/wfs/WFSRequestType.java:9]: Thanks! Yeah hopefully the enum helps with bookkeeping of extending this in the future as well.

File: tds/src/main/java/thredds/server/wfs/WFSController.java (hunk 3 above)
Comment (wchen329) [reply tds/src/main/java/thredds/server/wfs/WFSController.java]: Good catch (pun intended)! Yeah pretty much forgot about putting anything there. Perhaps logging with SLF4J and throwing another exception might be suitable here?

File: tds/src/main/java/thredds/server/wfs/WFSGetCapabilitiesWriter.java (hunk 4 above)
Comment (wchen329) [reply tds/src/main/java/thredds/server/wfs/WFSGetCapabilitiesWriter.java:15]: Yeah it is interesting to consider as well "states" that can be reached using the stateful design (i.e. its kind of funny to consider using finishXML() before startXML() hahaha and the sort of silly things that can happen). I have thought about using a non stateful design and at least it seemed less with the flow of the XML file (since as noted, it's often about writing one section than moving on to another due to its kind of tree'd structure). To tackle some of the issues with edge cases and just blatantly wrong usage of the writer, was thinking of enforcing ordering through exceptions or related on certain calls (that are akin to state transitions which might make it easier to test as well), though we can certainly consider de state-ifying it in MAP-104.

File: tds/src/main/java/thredds/server/wfs/WFSFeature.java (hunk 6 above)
Comment (wchen329) [reply tds/src/main/java/thredds/server/wfs/WFSFeature.java]: Yeah hmm....

At one point it was though it seems like it's not as of now? Maybe in the future to allow for CRS customizations. But removing it, thanks for finding that!

File: tds/src/main/java/thredds/server/wfs/WFSFeature.java (hunk 7 above)
Comment (wchen329) [reply tds/src/main/java/thredds/server/wfs/WFSFeature.java]: Done, make it immutable.

File: tds/src/main/java/thredds/server/wfs/WFSFeature.java (hunk 5 above)
Comment (wchen329) [reply tds/src/main/java/thredds/server/wfs/WFSFeature.java:9]: Thanks! Yeah seemed more approriate to make than making the forest-ish WFSGetCapabilities writer more forestish (as well as being good for modularity!).

**Comment by wchen329** [comment]
Thanks for all the feedback Carl! Will get another version pushed up ASAP.

File: tds/src/main/java/thredds/server/wfs/WFSController.java (hunk 1 above)
Comment (wchen329) [reply tds/src/main/java/thredds/server/wfs/WFSController.java:62]: Okay so this will be handled using the ows:Exception according to the standard. However, when GetFeature is implemented, using HTTP404 to handle non-existing datasets seems like the way to handle those things.

Comment (cschroedl-gov) [reply tds/src/main/java/thredds/server/wfs/WFSController.java:62]: Yesss! Good call! I had forgotten about the tendency to return Exception messages in XML.

File: tds/src/main/java/thredds/server/wfs/WFSLogger.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+/**
+ * A logger specifically used for logging events related to the WFS servlet.
+ * 
+ * @author wchen@usgs.gov
+ *
+ */
+public class WFSLogger {
+
+	static private final Logger wfsLog = LoggerFactory.getLogger(WFSController.class);
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/WFSLogger.java]: Good use of SLF4J!

Comment (cschroedl-gov) [reply tds/src/main/java/thredds/server/wfs/WFSLogger.java]: Ah, I missed this on the first pass...
SLF4J wants the current class to be passed to `LoggerFactory.getLogger`. SLF4J has [a warning](https://www.slf4j.org/codes.html#loggerNameMismatch) if the name of the Logger doesn't match the name of the class from which it is called.

File: tds/src/main/java/thredds/server/wfs/WFSController.java [hunk 9]
```diff
@@ -55,19 +54,22 @@ public void httpHandler(HttpServletRequest hsreq, HttpServletResponse hsres) {
...
-				wr.append(hsreq.getScheme() + "://" + hsreq.getServerName() + ":" + hsreq.getServerPort() + "/thredds/wfs");
+				WFSExceptionWriter owsExcept = new WFSExceptionWriter("WFS server error. REQUEST parameter is required.", "request", "MissingParameterValue", hsres);
+				owsExcept.write();
+				return;
 			}
 		}
 		
 		catch(IOException io) {
-			
+			WFSLogger.logError("ERROR: retrieval of writer failed.");
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/WFSController.java]: If you would like to log when an exception is caught, could logging methods be called directly on an unwrapped SLF4J Logger object? Doing so would permit logging the exception's stack trace (and any underlying exceptions' messages and stack traces) by passing the exception in the arguments. Example: `LOG.error("my custom message", io)`;

File: tds/src/main/java/thredds/server/wfs/WFSController.java [hunk 10]
```diff
@@ -55,19 +54,22 @@ public void httpHandler(HttpServletRequest hsreq, HttpServletResponse hsres) {
...
+			throw new RuntimeException("ERROR: retrieval of writer failed");
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/WFSController.java]: If you would like to throw an exception here, it might help to add `io` as a parameter to the `RuntimeException` constructor. This could help something further up the stack gain insight into the cause of the problem, and thus perform more helpful recovery. This could also help something further up the error-handling stack do better error reporting, by giving the error-handler access to the stack trace that caused the current problem.

File: tds/src/main/java/thredds/server/wfs/WFSController.java [hunk 11]
```diff
@@ -55,19 +54,22 @@ public void httpHandler(HttpServletRequest hsreq, HttpServletResponse hsres) {
...
 			else{
-				wr.append("ncWFS on THREDDS - ");
-				wr.append(hsreq.getScheme() + "://" + hsreq.getServerName() + ":" + hsreq.getServerPort() + "/thredds/wfs");
+				WFSExceptionWriter owsExcept = new WFSExceptionWriter("WFS server error. REQUEST parameter is required.", "request", "MissingParameterValue", hsres);
+				owsExcept.write();
+				return;
 			}
 		}
 		
 		catch(IOException io) {
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/WFSController.java]: Reading through log messages to diagnose a problem after it has happened is hard. It is easier if the same problem is only logged once. One widely-adopted convention that ensures a single source of truth in the event of a failure is to either rethrow an exception, or* to log it, but not both. Which do you think would be more appropriate here? Logging the exception or rethrowing the exception?

*"or" as in XOR.

File: tds/src/main/java/thredds/server/wfs/WFSLogger.java (hunk 8 above)
Comment (cschroedl-gov) [reply tds/src/main/java/thredds/server/wfs/WFSLogger.java]: Actually, I'm wondering now: is this class is needed? Would it suffice to instantiate an SLF4J `Logger` in each class that needs to do logging? `Logger`s are pretty cheap objects, so there's not much performance overhead in creating a new one per class. Adopting a one-logger-per-class practice permits frameworks like log4j to adjust the verbosity of logging on a per-class or per-package basis.

Remember when the dev VM's disk was eaten by a rogue logger on Jenkins? If the project in question used SLF4J and one-logger-per-class we could have resorted to shutting off logging for the one chatty class.

File: tds/src/main/java/thredds/server/wfs/WFSGetCapabilitiesWriter.java (hunk 4 above)
Comment (cschroedl-gov) [reply tds/src/main/java/thredds/server/wfs/WFSGetCapabilitiesWriter.java:15]: :) Very good points. Thanks for considering a stateless design. It's up to you!

File: tds/src/main/java/thredds/server/wfs/WFSLogger.java (hunk 8 above)
Comment (wchen329) [reply tds/src/main/java/thredds/server/wfs/WFSLogger.java]: The idea of using the WFS Logger specifically was to allow for a single SLF4J log for WFS related elements which is what is being done with DAP currently (see DapLog.class). Having a single "wfs" logger would be a good way to consolidate related messages (relating to wfs). Could certainly consider changing to a single "wfs" log or doing away with SLF4J in general for the time being considering your feedback below.

File: tds/src/main/java/thredds/server/wfs/WFSController.java (hunk 10 above)
Comment (wchen329) [reply tds/src/main/java/thredds/server/wfs/WFSController.java]: Ah yes, thanks for catching that Carl! Will pass it on through to the creation of Runtime Exception.

File: tds/src/main/java/thredds/server/wfs/WFSLogger.java (hunk 8 above)
Comment (wchen329) [reply tds/src/main/java/thredds/server/wfs/WFSLogger.java]: For sure, that is a benefit of having more granular control of the logging, but also brings up another point, whether or not logging is needed as of now. Given a small VM size, it could possibly cause denial of service by filling up the storage with SLF4J logs through malicious requests! Reverting to a single RuntimeException(...) as you posed earlier.

File: tds/src/main/java/thredds/server/wfs/WFSController.java (hunk 11 above)
Comment (wchen329) [reply tds/src/main/java/thredds/server/wfs/WFSController.java]: Yes, SLF4J shall be put on hold for the moment. Just will throw the single RuntimeException to compensate for this.

File: tds/src/main/java/thredds/server/wfs/WFSController.java (hunk 9 above)
Comment (wchen329) [reply tds/src/main/java/thredds/server/wfs/WFSController.java]: Mentioned in another comment that the wrapped WFSLogger was to provide a central log for all wfs related problems and information. Smaller more granular logs could be used, but the problem can also be solved by allowing for an Exception to be passed as a second argument. Regardless SLF4J stuff will be all removed in the time being in consideration of your comment below.

File: tds/src/main/java/thredds/server/wfs/WFSGetCapabilitiesWriter.java (hunk 4 above)
Comment (wchen329) [reply tds/src/main/java/thredds/server/wfs/WFSGetCapabilitiesWriter.java:15]: Thanks for looking through this pull request (and others as well) and providing suggestions on possible other methodologies that could be used!

File: tds/src/main/java/thredds/server/wfs/WFSController.java (hunk 9 above)
Comment (cschroedl-gov) [reply tds/src/main/java/thredds/server/wfs/WFSController.java]: :+1:

File: tds/src/main/java/thredds/server/wfs/WFSLogger.java (hunk 8 above)
Comment (cschroedl-gov) [reply tds/src/main/java/thredds/server/wfs/WFSLogger.java]: Cool! Makes Sense!

**Review by cschroedl-gov — APPROVED** [review]
